sled-storage/src/manager.rs

```diff
@@ -1499,27 +1498,9 @@ impl StorageManager {
             zoned,
             encryption_details,
             size_details,
-            id: Some(DatasetUuid::from_untyped_uuid(request.dataset_id)),
+            id: request.dataset_id,
             additional_options: None,
         })?;
-        // Ensure the dataset has a usable UUID.
```

Comment:
I'm still a little weirded out by this block of code. Was it strictly unnecessary before? If there's any chance it wasn't, should we keep it but inside an `if let Some(dataset_id) { ... }` block?

Reply:
I *think* it's a historical artifact -- I'd prefer to remove it if we can, because I think sled agent is plagued by too many of these "two ways to set a value" paths already, and I'd prefer to reduce it to one.

From 256dca4de51063689ea187c1c3f2a931b19fd3cf (2021) - https://github.com/oxidecomputer/omicron/blob/256dca4de51063689ea187c1c3f2a931b19fd3cf/sled-agent/src/storage_manager.rs#L232-L246 This is one of the earliest forms of "ensure filesystem, then set a value".

From ccc28fe4c857c08599f5d9d6eff6ecfcaa298eb6 (2023) - this was renamed, but still had the structure of "sled agent calls `Zfs::ensure_filesystem`, and **then** sets the UUID: https://github.com/oxidecomputer/omicron/blob/ccc28fe4c857c08599f5d9d6eff6ecfcaa298eb6/sled-agent/src/storage_manager.rs#L269-L298 

From de8bc93309f9eab8c431328065ea6f3d5bd0f34f (last week) -  `Zfs::ensure_filesystem` (later renamed to `ensure_dataset`) started being able to explicitly set properties, as part of the optimization to avoid calling `zfs set` individually for each property.

https://github.com/oxidecomputer/omicron/blob/de8bc93309f9eab8c431328065ea6f3d5bd0f34f/illumos-utils/src/zfs.rs#L550-L557

This was added recently, because (using dtrace) I noticed the callstack was basically:

```
zfs set <size properties , one at a time>
zfs get uuid
zfs set uuid
```

And by moving all properties into `Zfs::ensure_filesystem`, this could be a single call to `get/set`.

All this is to say:

- After de8bc93309f9eab8c431328065ea6f3d5bd0f34f, any property-managing code **after** calling `Zfs::ensure_filesystem` should not be necessary. If the UUID was supplied, it should have already been checked/set at this point.
- I'd prefer to remove this block of code, rather than diverge the code which mutates dataset UUIDs further.

Does that sound reasonable?

Reply:
Hah! Yes, I was doing the same digging and came to the same conclusion (i.e., #7236 made this code unnecessary). :+1: on the removal.